Add static and class subscripts

beccadax opened the pull request:
This PR lets you declare `static subscript`s or, in classes, `class subscript`s, removing an unnecessary asymmetry between different kinds of members.

This is just a quick Saturday prototype right now—it's nowhere near fully tested and I'm sure there are plenty of buried bodies I haven't discovered yet.

brentdax commented:
@swift-ci please test source compatibility

brentdax commented:
@swift-ci please build toolchain

brentdax commented:
@swift-ci please smoke test

brentdax commented:
@swift-ci please build toolchain

brentdax commented:
@swift-ci please build toolchain linux platform

brentdax commented:
@swift-ci please build toolchain linux platform

brentdax commented:
Rolled back work on metatype key paths, which are now out of scope. That work is parked in #23623.

brentdax commented:
@swift-ci please smoke test

brentdax commented:
@swift-ci please smoke test

brentdax commented:
@swift-ci please test compiler performance

brentdax commented:
@swift-ci please clean smoke test

brentdax commented:
@swift-ci please smoke test

brentdax commented:
# Summary for master full

**Unexpected test results, excluded stats for CoreStore, Tagged, ProcedureKit, Wordy, Deferred**

No regressions above thresholds

- [Debug-batch](#debug-batch-3d1b0a44b2c33e14)
    - [debug-batch brief](#debug-batch-brief-3d1b0a44b2c33e14)
    - [debug-batch detailed](#debug-batch-detailed-3d1b0a44b2c33e14)
- [Release](#release-3d1b0a44b2c33e14)
    - [release brief](#release-brief-3d1b0a44b2c33e14)
    - [release detailed](#release-detailed-3d1b0a44b2c33e14)



<a name="debug-batch-3d1b0a44b2c33e14">

# Debug-batch

<a name="debug-batch-brief-3d1b0a44b2c33e14">

## debug-batch brief
<details>
<summary>Regressed (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Improved (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (3)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
Frontend.NumInstructionsExecuted | 25,323,286,735,460 | 25,265,855,160,985 | -57,431,574,475 | -0.23%
LLVM.NumLLVMBytesOutput | 972,775,292 | 972,773,660 | -1,632 | -0.0%
time.swift-driver.wall | 2403.2s | 2403.1s | -133.0ms | -0.01%
</details>

<a name="debug-batch-detailed-3d1b0a44b2c33e14">

## debug-batch detailed
<details>
<summary>Regressed (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Improved (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (99)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
AST.NumASTBytesAllocated | 61,305,649,229 | 61,309,943,215 | 4,293,986 | 0.01%
AST.NumDecls | 80,786 | 80,786 | 0 | 0.0%
AST.NumDependencies | 160,472 | 160,466 | -6 | -0.0%
AST.NumImportedExternalDefinitions | 919,976 | 919,976 | 0 | 0.0%
AST.NumInfixOperators | 31,080 | 31,080 | 0 | 0.0%
AST.NumLinkLibraries | 0 | 0 | 0 | 0.0%
AST.NumLoadedModules | 196,321 | 196,321 | 0 | 0.0%
AST.NumLocalTypeDecls | 108 | 108 | 0 | 0.0%
AST.NumObjCMethods | 9,724 | 9,724 | 0 | 0.0%
AST.NumPostfixOperators | 18 | 18 | 0 | 0.0%
AST.NumPrecedenceGroups | 14,514 | 14,514 | 0 | 0.0%
AST.NumPrefixOperators | 70 | 70 | 0 | 0.0%
AST.NumReferencedDynamicNames | 92 | 92 | 0 | 0.0%
AST.NumReferencedMemberNames | 3,511,573 | 3,511,573 | 0 | 0.0%
AST.NumReferencedTopLevelNames | 271,022 | 271,022 | 0 | 0.0%
AST.NumSourceBuffers | 234,257 | 234,257 | 0 | 0.0%
AST.NumSourceLines | 2,432,547 | 2,432,547 | 0 | 0.0%
AST.NumSourceLinesPerSecond | 2,184,508 | 2,188,734 | 4,226 | 0.19%
AST.NumTotalClangImportedEntities | 3,596,501 | 3,599,818 | 3,317 | 0.09%
AST.NumUsedConformances | 226,386 | 226,386 | 0 | 0.0%
Driver.ChildrenMaxRSS | 107,199,283,200 | 107,239,983,104 | 40,699,904 | 0.04%
Driver.DriverDepCascadingDynamic | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingExternal | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingMember | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingNominal | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingTopLevel | 0 | 0 | 0 | 0.0%
Driver.DriverDepDynamic | 0 | 0 | 0 | 0.0%
Driver.DriverDepExternal | 0 | 0 | 0 | 0.0%
Driver.DriverDepMember | 0 | 0 | 0 | 0.0%
Driver.DriverDepNominal | 0 | 0 | 0 | 0.0%
Driver.DriverDepTopLevel | 0 | 0 | 0 | 0.0%
Driver.NumDriverJobsRun | 16,205 | 16,205 | 0 | 0.0%
Driver.NumDriverJobsSkipped | 0 | 0 | 0 | 0.0%
Driver.NumDriverPipePolls | 165,523 | 164,049 | -1,474 | -0.89%
Driver.NumDriverPipeReads | 177,350 | 176,204 | -1,146 | -0.65%
Driver.NumProcessFailures | 0 | 0 | 0 | 0.0%
Frontend.MaxMallocUsage | 587,635,049,752 | 587,873,188,904 | 238,139,152 | 0.04%
Frontend.NumInstructionsExecuted | 25,323,286,735,460 | 25,265,855,160,985 | -57,431,574,475 | -0.23%
Frontend.NumProcessFailures | 0 | 0 | 0 | 0.0%
IRModule.NumIRAliases | 105,540 | 105,540 | 0 | 0.0%
IRModule.NumIRBasicBlocks | 3,849,136 | 3,849,136 | 0 | 0.0%
IRModule.NumIRComdatSymbols | 0 | 0 | 0 | 0.0%
IRModule.NumIRFunctions | 1,804,505 | 1,804,505 | 0 | 0.0%
IRModule.NumIRGlobals | 1,857,058 | 1,857,058 | 0 | 0.0%
IRModule.NumIRIFuncs | 0 | 0 | 0 | 0.0%
IRModule.NumIRInsts | 47,452,775 | 47,452,775 | 0 | 0.0%
IRModule.NumIRNamedMetaData | 78,402 | 78,402 | 0 | 0.0%
IRModule.NumIRValueSymbols | 3,302,011 | 3,302,011 | 0 | 0.0%
LLVM.NumLLVMBytesOutput | 972,775,292 | 972,773,660 | -1,632 | -0.0%
Parse.NumFunctionsParsed | 142,585 | 142,585 | 0 | 0.0%
Parse.NumIterableDeclContextParsed | 709,593 | 709,593 | 0 | 0.0%
SILModule.NumSILGenDefaultWitnessTables | 0 | 0 | 0 | 0.0%
SILModule.NumSILGenFunctions | 932,961 | 932,961 | 0 | 0.0%
SILModule.NumSILGenGlobalVariables | 37,189 | 37,189 | 0 | 0.0%
SILModule.NumSILGenVtables | 10,694 | 10,694 | 0 | 0.0%
SILModule.NumSILGenWitnessTables | 39,362 | 39,362 | 0 | 0.0%
SILModule.NumSILOptDefaultWitnessTables | 0 | 0 | 0 | 0.0%
SILModule.NumSILOptFunctions | 1,326,972 | 1,326,972 | 0 | 0.0%
SILModule.NumSILOptGlobalVariables | 37,947 | 37,947 | 0 | 0.0%
SILModule.NumSILOptVtables | 17,207 | 17,207 | 0 | 0.0%
SILModule.NumSILOptWitnessTables | 86,613 | 86,613 | 0 | 0.0%
Sema.AccessLevelRequest | 2,171,313 | 2,171,511 | 198 | 0.01%
Sema.DefaultAndMaxAccessLevelRequest | 49,711 | 49,711 | 0 | 0.0%
Sema.DefaultTypeRequest | 315,674 | 315,674 | 0 | 0.0%
Sema.EnumRawTypeRequest | 15,117 | 15,117 | 0 | 0.0%
Sema.ExtendedNominalRequest | 3,197,579 | 3,200,871 | 3,292 | 0.1%
Sema.InheritedDeclsReferencedRequest | 3,892,938 | 3,902,559 | 9,621 | 0.25%
Sema.InheritedTypeRequest | 564,725 | 565,321 | 596 | 0.11%
Sema.IsDynamicRequest | 1,678,371 | 1,678,371 | 0 | 0.0%
Sema.IsObjCRequest | 1,435,594 | 1,435,542 | -52 | -0.0%
Sema.MangleLocalTypeDeclRequest | 216 | 216 | 0 | 0.0%
Sema.NamedLazyMemberLoadFailureCount | 17,007 | 17,017 | 10 | 0.06%
Sema.NamedLazyMemberLoadSuccessCount | 16,443,501 | 16,450,429 | 6,928 | 0.04%
Sema.NominalTypeLookupDirectCount | 27,459,077 | 27,467,332 | 8,255 | 0.03%
Sema.NumConformancesDeserialized | 6,180,248 | 6,190,799 | 10,551 | 0.17%
Sema.NumConstraintScopes | 14,816,710 | 14,818,660 | 1,950 | 0.01%
Sema.NumConstraintsConsideredForEdgeContraction | 41,224,999 | 41,225,574 | 575 | 0.0%
Sema.NumDeclsDeserialized | 43,889,207 | 43,952,520 | 63,313 | 0.14%
Sema.NumDeclsFinalized | 1,499,521 | 1,499,521 | 0 | 0.0%
Sema.NumDeclsTypechecked | 895,861 | 895,861 | 0 | 0.0%
Sema.NumDeclsValidated | 1,814,194 | 1,814,194 | 0 | 0.0%
Sema.NumFunctionsTypechecked | 848,567 | 848,567 | 0 | 0.0%
Sema.NumGenericSignatureBuilders | 1,011,622 | 1,012,806 | 1,184 | 0.12%
Sema.NumLazyGenericEnvironments | 8,672,415 | 8,677,630 | 5,215 | 0.06%
Sema.NumLazyGenericEnvironmentsLoaded | 188,247 | 188,435 | 188 | 0.1%
Sema.NumLazyIterableDeclContexts | 5,730,093 | 5,734,790 | 4,697 | 0.08%
Sema.NumLeafScopes | 9,619,668 | 9,621,124 | 1,456 | 0.02%
Sema.NumTypesDeserialized | 14,308,589 | 14,323,157 | 14,568 | 0.1%
Sema.NumTypesValidated | 1,227,708 | 1,227,706 | -2 | -0.0%
Sema.NumUnloadedLazyIterableDeclContexts | 3,608,604 | 3,610,542 | 1,938 | 0.05%
Sema.OverriddenDeclsRequest | 7,276,296 | 7,286,516 | 10,220 | 0.14%
Sema.RequirementRequest | 59,884 | 59,884 | 0 | 0.0%
Sema.SelfBoundsFromWhereClauseRequest | 5,653,243 | 5,667,892 | 14,649 | 0.26%
Sema.SetterAccessLevelRequest | 126,612 | 126,612 | 0 | 0.0%
Sema.SuperclassDeclRequest | 59,046 | 59,301 | 255 | 0.43%
Sema.SuperclassTypeRequest | 28,517 | 28,517 | 0 | 0.0%
Sema.TypeDeclsFromWhereClauseRequest | 28,021 | 28,021 | 0 | 0.0%
Sema.USRGenerationRequest | 12,449,698 | 12,474,064 | 24,366 | 0.2%
Sema.UnderlyingTypeDeclsReferencedRequest | 134,261 | 134,421 | 160 | 0.12%
</details>

<a name="release-3d1b0a44b2c33e14">

# Release

<a name="release-brief-3d1b0a44b2c33e14">

## release brief
<details>
<summary>Regressed (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Improved (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (3)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
Frontend.NumInstructionsExecuted | 25,575,835,994,835 | 25,528,786,137,206 | -47,049,857,629 | -0.18%
LLVM.NumLLVMBytesOutput | 779,847,214 | 779,843,818 | -3,396 | -0.0%
time.swift-driver.wall | 4061.9s | 4062.6s | 665.5ms | 0.02%
</details>

<a name="release-detailed-3d1b0a44b2c33e14">

## release detailed
<details>
<summary>Regressed (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Improved (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (23)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
AST.NumImportedExternalDefinitions | 216,559 | 216,559 | 0 | 0.0%
AST.NumLoadedModules | 16,492 | 16,492 | 0 | 0.0%
AST.NumTotalClangImportedEntities | 736,667 | 736,667 | 0 | 0.0%
AST.NumUsedConformances | 227,331 | 227,331 | 0 | 0.0%
IRModule.NumIRBasicBlocks | 3,234,121 | 3,234,121 | 0 | 0.0%
IRModule.NumIRFunctions | 1,486,895 | 1,486,895 | 0 | 0.0%
IRModule.NumIRGlobals | 1,622,639 | 1,622,639 | 0 | 0.0%
IRModule.NumIRInsts | 29,381,129 | 29,381,129 | 0 | 0.0%
IRModule.NumIRValueSymbols | 2,896,376 | 2,896,376 | 0 | 0.0%
LLVM.NumLLVMBytesOutput | 779,847,214 | 779,843,818 | -3,396 | -0.0%
SILModule.NumSILGenFunctions | 653,152 | 653,152 | 0 | 0.0%
SILModule.NumSILOptFunctions | 889,962 | 889,962 | 0 | 0.0%
Sema.NumConformancesDeserialized | 2,233,184 | 2,233,184 | 0 | 0.0%
Sema.NumConstraintScopes | 13,401,956 | 13,401,956 | 0 | 0.0%
Sema.NumDeclsDeserialized | 6,021,369 | 6,021,369 | 0 | 0.0%
Sema.NumDeclsValidated | 1,038,655 | 1,038,655 | 0 | 0.0%
Sema.NumFunctionsTypechecked | 428,737 | 428,737 | 0 | 0.0%
Sema.NumGenericSignatureBuilders | 190,265 | 190,265 | 0 | 0.0%
Sema.NumLazyGenericEnvironments | 1,245,248 | 1,245,248 | 0 | 0.0%
Sema.NumLazyGenericEnvironmentsLoaded | 21,173 | 21,173 | 0 | 0.0%
Sema.NumLazyIterableDeclContexts | 765,629 | 765,629 | 0 | 0.0%
Sema.NumTypesDeserialized | 3,186,576 | 3,186,576 | 0 | 0.0%
Sema.NumTypesValidated | 616,421 | 616,421 | 0 | 0.0%
</details>

brentdax commented:
@swift-ci please test

brentdax commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please build toolchain

DougGregor commented:
@swift-ci please build toolchain

brentdax commented:
@swift-ci please smoke test

brentdax commented:
@swift-ci please smoke test

brentdax commented:
@swift-ci please smoke test compiler performance

brentdax commented:
@swift-ci please build toolchain

brentdax commented:
@swift-ci please build toolchain

beccadax committed:
Fix description of static subscripts in errors

beccadax committed:
Make sure we don’t support @objc class subscripts

Obj-C subscripts don’t work on class objects, but you can declare class methods with the appropriate names. We don’t want to half-support this. Emit an error if you try to write “@objc class subscript” and test that we don’t import the methods as subscripts.

beccadax committed:
Permit subscripting types without using .self

beccadax committed:
Infer final on static subscripts

beccadax committed:
Add parallel tests for static subscripts

This commit modifies various subscript-related test files to add static subscript equivalents of existing tests.

beccadax committed:
Allow the declaration of static subscripts

In this commit, MyStruct.self[0] parses and typechecks but the solution doesn’t apply correctly. MyStruct[0] gets diagnosed as an error.

beccadax committed:
Correctly apply typechecking solutions with subscripts on type instances

brentdax commented:
Now that many of the underlying changes have been merged to master in #23724, I've rebased the remaining commits and cleaned up the version history a bit.

brentdax commented:
@swift-ci please test source compatibility

brentdax commented:
@swift-ci please test compiler performance

beccadax committed:
[ASTPrinter] Print staticness of subscripts

brentdax commented:
@swift-ci please test compiler performance

jrose-apple commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2929,13 +2929,10 @@ Parser::parseDecl(ParseDeclOptions Flags,
-    DeclResult = parseDeclSubscript(Flags, Attributes, Entries);
+    DeclResult = parseDeclSubscript(StaticLoc, StaticSpelling, Flags,
+                                    Attributes, Entries);
+    StaticLoc = SourceLoc(); // we handled static if present.
```
Do you need to clear this?

brentdax replied:
Yes. If StaticLoc is not cleared by the end of the function, it assumes you applied `static` to a declaration that doesn't support it and diagnoses an error.

jrose-apple commented on test/attr/attr_override.swift:
```diff
@@ -40,7 +40,7 @@ class A {
   var v10: Int { return 5 } // expected-note{{attempt to override property here}}
 
-  subscript (i: Int) -> String { // expected-note{{potential overridden subscript 'subscript(_:)' here}}
+  subscript (i: Int) -> String { // expected-note{{potential overridden subscript 'subscript(_:)' here}} expected-note{{potential overridden subscript 'subscript(_:)' here}}
```
Hm, shouldn't we only be showing overridables with the same static-ness?

brentdax replied:
You're right—we don't show that for class functions (although I'm not actually sure where we test that). I'll look into this.

brentdax replied:
No, it looks like I was right. This note was generated by the new instance subscript on line 181, which was added to test that the compiler doesn't think an instance subscript can override a class subscript with the same signature. But I could avoid these notes by tweaking the test a little.

brentdax replied:
Improving that test revealed [a subscript override-checking bug unrelated to static subscripts](https://bugs.swift.org/browse/SR-10323). I've pushed the modified test with FIXMEs referring to this bug, but I'll amend it if I end up fixing the new bug first.

jrose-apple commented on test/decl/subscript/static.swift:
```diff
@@ -3,10 +3,34 @@
+  override static subscript(_ i: Int) -> String { return "DerivedGood" }
+}
+
+print(DerivedGood[0])
```
Can you have a test where you subscript a subclass but it *doesn't* override the subscript implementation?

brentdax replied:
I suppose so.

(Actually, I just realized this test is wrong—the base class subscript should be implicitly `final`. I'll have to figure out what's happening there...)

brentdax replied:
I've added some override tests to test/Interpreter/generic_subscript_static.swift, including one that tests a subscript that doesn't appear in the base class. I've also added some SILGen tests to make sure we handle class subscript overrides the same way as instance subscript overrides.

jrose-apple commented on lib/Sema/TypeCheckDeclObjC.cpp:
```diff
@@ -827,6 +827,16 @@ bool swift::isRepresentableInObjC(const SubscriptDecl *SD, ObjCReason Reason) {
   if (checkObjCInForeignClassContext(SD, Reason))
     return false;
 
+  // Obj-C doesn't support class subscripts.
```
Nitpick: "Objective-C" but "ObjC".

jrose-apple commented on test/Interpreter/generic_subscript_static.swift:
```diff
@@ -0,0 +1,142 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
```
Nitpick: 2019

brentdax replied:
Updated to 2014 - 2019, since these tests are basically modified versions of another test file dating back that far.

jrose-apple reviewed (commented):
Nice work! Only small comments from me.

beccadax committed:
[NFC] Pick a few nits for Jordan

brentdax commented:
@swift-ci please smoke test

brentdax commented:
@swift-ci please smoke test macOS platform

brentdax commented:
@swift-ci please smoke test

brentdax commented:
@swift-ci please smoke test

johnfairh commented:
@brentdax are you going to make a SourceKit change to go with this, to make static and instance subscripts look different?

brentdax commented:
@johnfairh Not me personally, but the SourceKit team is scheduled to work on it.

johnfairh commented:
Great, thanks.
